server/app/durablejobs/RecurringJobScheduler.java

```diff
@@ -0,0 +1,111 @@
+
+  @Inject
+  public RecurringJobScheduler(
+      Clock clock,
+      DurableJobRegistry durableJobRegistry,
+      PersistedDurableJobRepository persistedDurableJobRepository) {
+    this.clock = Preconditions.checkNotNull(clock);
+    this.durableJobRegistry = Preconditions.checkNotNull(durableJobRegistry);
+    this.persistedDurableJobRepository = Preconditions.checkNotNull(persistedDurableJobRepository);
+  }
+
+  /**
+   * Checks that each recurring job in {@link DurableJobRegistry} has a {@link PersistedDurableJob}
+   * scheduled sometime in the future and schedules one if not.
+   *
+   * <p>{@code synchronized} to avoid overlapping executions within the same server.
+   */
+  public synchronized void scheduleRecurringJobs() {
+    for (DurableJobRegistry.RegisteredJob registeredJob : durableJobRegistry.getRecurringJobs()) {
+      if (registeredJob.getRecurringJobExecutionTimeResolver().isEmpty()) {
+        LOGGER.error(
+            "JobScheduler_SchedulingError No RecurringJobExecutionTimeResolver registered with {}",
+            registeredJob.getJobName());
+        continue;
+      }
+
+      Instant executionTime =
+          registeredJob.getRecurringJobExecutionTimeResolver().get().resolveExecutionTime(clock);
+      boolean jobAlreadyScheduled =
+          persistedDurableJobRepository
+              .findScheduledJob(registeredJob.getJobName().getJobNameString(), executionTime)
+              .isPresent();
+
+      if (!jobAlreadyScheduled) {
+        PersistedDurableJob newJob =
+            new PersistedDurableJob(registeredJob.getJobName().getJobNameString(), executionTime);
+        try {
+          tryScheduleRecurringJob(newJob, SCHEDULER_ATTEMPTS);
+        } catch (InterruptedException e) {
+          throw new RuntimeException(e);
+        }
+      }
+    }
+  }
+
+  private synchronized void tryScheduleRecurringJob(
+      PersistedDurableJob newJob, int remainingAttempts) throws InterruptedException {
+    try (Transaction transaction = database.beginTransaction(TxIsolation.SERIALIZABLE)) {
+
+      // Re-fetch upon each attempt so the transaction prevents duplicates.
+      if (!persistedDurableJobRepository
+          .findScheduledJob(newJob.getJobName(), newJob.getExecutionTime())
+          .isPresent()) {
+        newJob.save(transaction);
+        transaction.commit();
+      }
+    } catch (OptimisticLockException e) {
+      if (remainingAttempts > 0) {
+        LOGGER.warn(
+            "JobScheduler_SchedulingError OptimisticLockException scheduling {} with {} remaining"
+                + " attempts",
+            newJob.getJobName(),
+            remainingAttempts);
+
+        // Sleep for up to 100ms in effort to avoid transaction contention with other servers
+        // that may be attempting to schedule the job at the same time.
+        Thread.sleep(new Random().nextInt(/* bound= */ 100));
+        tryScheduleRecurringJob(newJob, remainingAttempts--);
```

Comment:
This is just my unfamiliarity with this "try with resources" thing, but from reading https://ebean.io/docs/transactions/, I'm assuming the transaction is closed immediately when the exception is thrown, not after it's handled, and we won't potentially get several open transactions at once?

Reply:
> I'm assuming the transaction is closed immediately when the exception is thrown, not after it's handled

yes. the transaction is closed once the `try` block exits, so after the exception is thrown but before it's handled with `catch`